### PR TITLE
Use `ConstArgKind::Path` for all single-segment paths, not just params under `min_generic_const_args`

### PR DESCRIPTION
**Pull request by camelid**
r? @BoxyUwU

edit by @BoxyUwU:

This PR introduces a `min_generic_const_args` feature gate and implements some preliminary work for it, representing all const arguments that are single segment paths as `ConstArg::Path` instead of only those that resolve to a const generic parameter. There are a few bits of follow up work after this lands:
- Figure out how to represent `Foo<{ STATIC }>`
- Figure out how to evaluate `Foo<{ EnumVariantConstructor }>`
- Make param env normalization handle non-anon-consts
- Move `try_from_lit` and `from_anon_const` to hir ty lowering too

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -2299,7 +2299,11 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         span: Span,
     ) -> &'hir hir::ConstArg<'hir> {
         let ct_kind = match res {
-            Res::Def(DefKind::ConstParam, _) => {
+            // FIXME(min_generic_const_args): only allow one-segment const paths for now
+            Res::Def(
```
Comment (BoxyUwU) [inline compiler/rustc_ast_lowering/src/lib.rs:2303]: We shouldn't match on the res as all single segment path resolutions should start being handled in hir ty lowering instead.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -2376,8 +2380,15 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         debug!("res={:?}", maybe_res);
         // FIXME(min_generic_const_args): for now we only lower params to ConstArgKind::Path
         if let Some(res) = maybe_res
-            && let Res::Def(DefKind::ConstParam, _) = res
             && let ExprKind::Path(qself, path) = &expr.kind
+            // FIXME(min_generic_const_args): only allow one-segment const paths for now
+            && let Res::Def(
```
Comment (BoxyUwU) [inline compiler/rustc_ast_lowering/src/lib.rs:2385]: same here

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 3]
```diff
@@ -1988,13 +2091,6 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     }
 
     /// Lower a type from the HIR to our internal notion of a type given some extra data for diagnostics.
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2093]:
```suggestion
     /// Lower a type from the HIR to our internal notion of a type.
```

(preexisting) this too

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 4]
```diff
@@ -1959,23 +1974,111 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+        let tcx = self.tcx();
+
+        if let FeedConstTy::Param(param_def_id) = feed
+            && let hir::ConstArgKind::Anon(anon) = &const_arg.kind
+        {
+            tcx.feed_anon_const_type(anon.def_id, tcx.type_of(param_def_id));
+        }
+
+        match const_arg.kind {
+            hir::ConstArgKind::Path(qpath) => self.lower_const_arg_path(qpath, const_arg.hir_id),
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2014]: I think for symmetry with the type lowering methods we should match on the `qpath` here, this is effectively just me asking you to remove `lower_const_arg_path` and inline it into this function I think. If you look at `fn lower_ty` you'll see what I mean about the symmetry thing, that function matches on the specific kind of `TyKind::Path` it has then dispatches to different functions based off it. There's no equivalent to `lower_const-arg_path` for types.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 5]
```diff
@@ -1959,23 +1974,111 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+            hir::ConstArgKind::Anon(anon) => Const::from_anon_const(tcx, anon.def_id),
+        }
+    }
+
+    /// Lower a const path to a [`Const`].
+    fn lower_const_arg_path(&self, qpath: hir::QPath<'tcx>, hir_id: HirId) -> Const<'tcx> {
+        let tcx = self.tcx();
+
+        match qpath {
+            hir::QPath::Resolved(_, &hir::Path { res: Res::Def(DefKind::ConstParam, did), .. }) => {
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2024]: This arm can just be removed as `lower_const_path_resolved` handles this

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 6]
```diff
@@ -1959,23 +1974,111 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+                self.lower_const_param(did, hir_id)
+            }
+            hir::QPath::Resolved(
+                _,
+                &hir::Path { res: Res::Def(DefKind::Fn | DefKind::AssocFn, _), .. },
+            ) => ty::Const::new_error_with_message(
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2030]: This arm should be moved into `lower_const_path_resolved`

**Review by BoxyUwU — COMMENTED** [review]
:sparkles: thanks for working on this, excited to be able to add the `min_generic_const_args` feature gate soon after all of these refactorings :sweat_smile:

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 7]
```diff
@@ -1959,23 +1974,111 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+                    GenericsArgsErrExtend::None,
+                );
+                let args = self.lower_generic_args_of_path_segment(
+                    span,
+                    did,
+                    path.segments.last().unwrap(),
+                );
+                ty::Const::new_unevaluated(tcx, ty::UnevaluatedConst::new(did, args))
+            }
+            _ => Const::new_error(
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2078]: Can you exhaustively match on every `DefKind` instead of using a wildcard pattern. It'd be easier to look at this code and check whether we're handling every case we should be if I don't have to open a separate tab to look at the definition of `DefKind` then cross reference it with the match arms you wrote :3

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 8]
```diff
@@ -43,6 +43,7 @@ fn eval_body_using_ecx<'tcx, R: InterpretationResult<'tcx>>(
                     | DefKind::AnonConst
                     | DefKind::InlineConst
                     | DefKind::AssocConst
+                    | DefKind::Ctor(_, CtorKind::Const)
```
[the remark this answers is not in the document]
Comment (camelid) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:46]: Yeah, this ICEs with

```
thread 'rustc' panicked at compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs:243:1:
DefId(20:5 ~ xcrate_const_ctor_a[1188]::Foo::{constructor#0}) does not have a "mir_for_ctfe"
stack backtrace:
```

File: compiler/rustc_hir_analysis/src/lib.rs [hunk 9]
```diff
@@ -227,3 +229,13 @@ pub fn lower_ty<'tcx>(tcx: TyCtxt<'tcx>, hir_ty: &hir::Ty<'tcx>) -> Ty<'tcx> {
     let env_def_id = tcx.hir().get_parent_item(hir_ty.hir_id);
     collect::ItemCtxt::new(tcx, env_def_id.def_id).lower_ty(hir_ty)
 }
+
+/// This is for rustdoc and clippy.
+pub fn lower_const_arg<'tcx>(
```
[the remark this answers is not in the document]
Comment (camelid) [reply compiler/rustc_hir_analysis/src/lib.rs:228]: The whole story of consts in rustdoc is a mess (see #83035). The TLDR is that we use both `hir::Const` and `ty::Const` and we basically need to use both. It's an open question whether/in which cases rustdoc should show the original const expression (`hir`) or the evaluated version (`ty`). There's advantages to both. So in the middle of processing some `hir` nodes, rustdoc may need to lower a hir const to a ty const. I don't think we'll be able to get rid of that anytime soon.

You're right that clippy doesn't need it though; I think it did when I originally wrote this code but was since modified to no longer need it.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs (hunk 5 above)
Comment (camelid) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2068]: Whoops, good catch.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs (hunk 7 above)
Comment (camelid) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs]: Done -- pretty long though haha.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 10]
```diff
@@ -2003,23 +2018,138 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+                tcx,
+                span,
+                "fn items cannot be used as const args",
+            ),
+
+            // Exhaustive match to be clear about what exactly we're considering to be
+            // an invalid Res for a const path.
+            Res::Def(
+                DefKind::Mod
+                | DefKind::Static { .. }
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2113]: `Static` ought to be reachable, so we should emit an error instead of ICEing here. Do we already have a test for `Foo<{ MY_STATIC }>`? Can you check if this actually *does* ICE under your PR, if it doesn't I'd be curious to find out why

You might need the feature gate for #119618 to make it ICE

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 11]
```diff
@@ -2003,23 +2018,138 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+                    GenericsArgsErrExtend::None,
+                );
+                let args = self.lower_generic_args_of_path_segment(
+                    span,
+                    did,
+                    path.segments.last().unwrap(),
+                );
+                ty::Const::new_unevaluated(tcx, ty::UnevaluatedConst::new(did, args))
+            }
+            Res::Def(DefKind::Fn | DefKind::AssocFn, _) => ty::Const::new_error_with_message(
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2103]: I don't *think* `AssocFn` is able to appear here until we start supporting multi-segment paths.

I think one day we will wind up supporting `DefKind::Fn` in const generics so could you add a `FIXME(const_generics)` here to create a real `ty::Const` here instead of delaying an ICE. I'm not sure how we'll actually want to represent this yet since `ConstKind::Unevaluated` doesn't seem right ^^

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 12]
```diff
@@ -2003,23 +2018,138 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+                _,
+            )
+            | Res::PrimTy(_)
+            | Res::SelfTyParam { .. }
+            | Res::SelfTyAlias { .. }
+            | Res::SelfCtor(_)
+            | Res::Local(_)
+            | Res::ToolMod
+            | Res::NonMacroAttr(_)
+            | Res::Err => Const::new_error(
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2149]: `new_error_with_message`?

File: compiler/rustc_hir_analysis/src/lib.rs [hunk 13]
```diff
@@ -221,3 +223,13 @@ pub fn lower_ty<'tcx>(tcx: TyCtxt<'tcx>, hir_ty: &hir::Ty<'tcx>) -> Ty<'tcx> {
     let env_def_id = tcx.hir().get_parent_item(hir_ty.hir_id);
     collect::ItemCtxt::new(tcx, env_def_id.def_id).lower_ty(hir_ty)
 }
+
+/// This is for rustdoc.
+pub fn lower_const_arg<'tcx>(
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/lib.rs:228]:
```suggestion
FIXME(const_generics): having special methods for rustdoc in `rustc_hir_analysis` is cursed
pub fn lower_const_arg_for_rustdoc<'tcx>(
```
seems easy to miss the doc comment now that I think about it xd

File: compiler/rustc_middle/src/ty/consts.rs [hunk 14]
```diff
@@ -184,46 +185,7 @@ impl<'tcx> rustc_type_ir::inherent::Const<TyCtxt<'tcx>> for Const<'tcx> {
...
-                // FIXME(min_generic_const_args): for now only params are lowered to ConstArgKind::Path
-                Self::from_param(tcx, qpath, const_arg.hir_id)
-            }
-            hir::ConstArgKind::Anon(anon) => Self::from_anon_const(tcx, anon.def_id),
-        }
-    }
-
     /// Literals and const generic parameters are eagerly converted to a constant, everything else
     /// becomes `Unevaluated`.
     #[instrument(skip(tcx), level = "debug")]
```
Comment (BoxyUwU) [inline compiler/rustc_middle/src/ty/consts.rs:191]: If we're going to punt on this to a follow up PR can you add a `FIXME(const_generics)` so we don't forget

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs (hunk 10 above)
Comment (camelid) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2113]: You're right, it does ICE. Though actually I think we need to allow it because at least on nightly, the following code passes:

```rust
pub static STATIC: u32 = 0;
pub struct Foo<const N: u32>;
pub const FOO: Foo<{STATIC}> = Foo;

fn main() {}
```

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs (hunk 11 above)
Comment (camelid) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2128]: I feel like it's fine to include `AssocFn` here for now though, no? Since we will want to handle it pretty much the same as `Fn` anyway once we do support them.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 15]
```diff
@@ -2025,23 +2040,138 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     ///
     /// Early-bound const parameters get lowered to [`ty::ConstKind::Param`]
     /// and late-bound ones to [`ty::ConstKind::Bound`].
-    pub(crate) fn lower_const_param(&self, hir_id: HirId) -> Const<'tcx> {
```
Comment (BoxyUwU) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2028]: how come this was changed. I think it would be nicer to keep it the same as it was since `lower_ty_param` also doesn't take a `DefId`

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs (hunk 8 above)
Comment (BoxyUwU) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:46]: since this is all just feature gated now can u remove this change and we can figure out how to handle these constructors at a later time so we don't have to worry about it rn

File: tests/ui/stable-mir-print/operands.stdout [hunk 16]
```diff
@@ -254,6 +254,13 @@ fn closures::{closure#0}(_1: {closure@$DIR/operands.rs:47:5: 47:19}, _2: bool) -
         return;
     }
 }
+fn Ctors::Unit() -> Ctors {
```
Comment (BoxyUwU) [inline tests/ui/stable-mir-print/operands.stdout:257]: you'll need to run `x test tests/ui/stable-mir-print --bless` and undo this change to stdout

File: tests/crashes/auxiliary/aux132985.rs [hunk 17]
```diff
@@ -0,0 +1,6 @@
+#![feature(adt_const_params)]
```
Comment (BoxyUwU) [inline tests/crashes/auxiliary/aux132985.rs:1]: apparently `crashes` tests dont support aux build, can either wait for #133013 or just remove this test and land without it since we're tracking it in an issue anyway :woman_shrugging:

File: compiler/rustc_trait_selection/src/traits/mod.rs [hunk 18]
```diff
@@ -372,9 +371,7 @@ pub fn normalize_param_env_or_error<'tcx>(
                     // should actually be okay since without `feature(generic_const_exprs)` the only
                     // const arguments that have a non-empty param env are array repeat counts. These
                     // do not appear in the type system though.
-                    if let ty::ConstKind::Unevaluated(uv) = c.kind()
-                        && self.0.def_kind(uv.def) == DefKind::AnonConst
```
Comment (BoxyUwU) [inline compiler/rustc_trait_selection/src/traits/mod.rs:376]: dont need to remove this now that we're feature gating stuff

**Comment by BoxyUwU** [comment]
also you'll need to squash all these commits before merging

**Comment by BoxyUwU** [comment]
rebased and squashed commits for you as well as blessing the stable mir test now that those changes have been reverted. we can follow up on everything else later. Having the feature gate will be super nice thanks :3

@bors r+
